tests/test_ops.py

```diff
@@ -0,0 +1,294 @@
+    df["chrom_2"] = df["chrom_1"].values
+    return df
+
+
+chroms = ["chr12", "chrX"]
+
+
+def mock_bioframe(num_entries=100):
+    pos = np.random.randint(1, 1e7, size=(num_entries, 2))
+    df = pd.DataFrame()
+    df["chrom"] = np.random.choice(chroms, num_entries)
+    df["start"] = np.min(pos, axis=1)
+    df["end"] = np.max(pos, axis=1)
+    df.sort_values(["chrom", "start"], inplace=True)
+    return df
+
+
+############# tests #####################
+
+
+def test_expand():
+    fake_bioframe = pd.DataFrame(
+        {"chrom": ["chr1", "chr1", "chr2"], "start": [1, 50, 100], "end": [5, 55, 200]}
+    )
+    fake_chromsizes = {"chr1": 60, "chr2": 300}
+    expand_bp = 10
+    fake_expanded = bioframe.expand(fake_bioframe.copy(), expand_bp, fake_chromsizes)
+    print(fake_expanded)
+    assert fake_expanded.iloc[0].start == 0  # don't expand below zero
+    assert (
+        fake_expanded.iloc[1].end == fake_chromsizes["chr1"]
+    )  # don't expand above chromsize
+    assert (
+        fake_expanded.iloc[2].end == fake_bioframe.iloc[2].end + expand_bp
+    )  # expand end normally
+    assert (
+        fake_expanded.iloc[2].start == fake_bioframe.iloc[2].start - expand_bp
+    )  # expand start normally
+
+
+def test_overlap():
+    ### note does not test overlap_start or overlap_end columns of bioframe.overlap
+    df1 = mock_bioframe()
+    df2 = mock_bioframe()
+    assert df1.equals(df2) == False
+    p1 = bioframe_to_pyranges(df1)
+    p2 = bioframe_to_pyranges(df2)
+    pp = pyranges_overlap_to_bioframe(p1.join(p2))[
+        ["chrom_1", "start_1", "end_1", "chrom_2", "start_2", "end_2"]
+    ]
+    bb = bioframe.overlap(df1, df2)[
+        ["chrom_1", "start_1", "end_1", "chrom_2", "start_2", "end_2"]
+    ]
+    pd.testing.assert_frame_equal(bb, pp, check_dtype=False, check_exact=True)
+    print("overlap elements agree")
+
+
+def test_cluster():
+    df1 = pd.DataFrame(
+        [["chr1", 1, 5], ["chr1", 3, 8], ["chr1", 8, 10], ["chr1", 12, 14],],
+        columns=["chrom", "start", "end"],
+    )
+    df_annotated = bioframe.cluster(df1)
+    assert (
+        df_annotated["cluster"].values == np.array([0, 0, 0, 1])
+    ).all()  # the last interval does not overlap the first three
+    df_annotated = bioframe.cluster(df1, min_dist=2)
+    assert (
+        df_annotated["cluster"].values == np.array([0, 0, 0, 0])
+    ).all()  # all intervals part of the same cluster
+
+    df_annotated = bioframe.cluster(df1, min_dist=None)
+    assert (
+        df_annotated["cluster"].values == np.array([0, 0, 1, 2])
+    ).all()  # adjacent intervals not clustered
+
+    df1.iloc[0, 0] = "chrX"
+    df_annotated = bioframe.cluster(df1)
+    assert (
+        df_annotated["cluster"].values == np.array([2, 0, 0, 1])
+    ).all()  # do not cluster intervals across chromosomes
+
+    # test consistency with pyranges (which automatically sorts df upon creation and uses 1-based indexing for clusters)
+    assert (
```

Comment:
oh, i see, so you test most functions against pyranges... I wonder if, in the future, it'd make sense to split these tests?